nexus/src/db/datastore/volume.rs

```diff
@@ -335,6 +335,99 @@ impl DataStore {
             })
     }
 
+    /// Create new UUIDs for the volume construction request layers
+    pub fn randomize_ids(
+        vcr: &VolumeConstructionRequest,
+    ) -> anyhow::Result<VolumeConstructionRequest> {
+        match vcr {
+            VolumeConstructionRequest::Volume {
+                id: _,
+                block_size,
+                sub_volumes,
+                read_only_parent,
+            } => Ok(VolumeConstructionRequest::Volume {
+                id: Uuid::new_v4(),
+                block_size: *block_size,
+                sub_volumes: sub_volumes
+                    .iter()
+                    .map(
+                        |subvol| -> anyhow::Result<VolumeConstructionRequest> {
+                            Self::randomize_ids(&subvol)
+                        },
+                    )
+                    .collect::<anyhow::Result<Vec<VolumeConstructionRequest>>>(
+                    )?,
+                read_only_parent: if let Some(read_only_parent) =
+                    read_only_parent
+                {
+                    Some(Box::new(Self::randomize_ids(read_only_parent)?))
+                } else {
+                    None
+                },
+            }),
+
+            VolumeConstructionRequest::Url { id: _, block_size, url } => {
+                Ok(VolumeConstructionRequest::Url {
+                    id: Uuid::new_v4(),
+                    block_size: *block_size,
+                    url: url.clone(),
+                })
+            }
+
+            VolumeConstructionRequest::Region {
+                block_size,
+                blocks_per_extent,
+                extent_count,
+                opts,
+                gen,
+            } => {
+                let mut opts = opts.clone();
+                opts.id = Uuid::new_v4();
+
+                Ok(VolumeConstructionRequest::Region {
+                    block_size: *block_size,
+                    blocks_per_extent: *blocks_per_extent,
+                    extent_count: *extent_count,
+                    opts,
+                    gen: *gen,
+                })
+            }
+
+            VolumeConstructionRequest::File { id: _, block_size, path } => {
+                Ok(VolumeConstructionRequest::File {
+                    id: Uuid::new_v4(),
+                    block_size: *block_size,
+                    path: path.clone(),
+                })
+            }
+        }
+    }
+
+    /// Checkout a copy of the Volume from the database using `volume_checkout`,
+    /// then randomize the UUIDs in the construction request. Because this is a
+    /// new volume, it is immediately passed to `volume_create` so that the
+    /// accounting for Crucible resources stays correct.
+    pub async fn volume_checkout_randomize_ids(
```

Comment:
Excuse my ignorance, but why do we need new/different volume IDs?

Reply:
This is called when creating an image out of a snapshot. Without an extra volume, the accounting for the underlying crucible resources associated with that snapshot won't be incremented, and if the user deletes that snapshot the underlying crucible resources will be freed. With this new volume, the user can delete the snapshot and still make disks out of this image.